packages/Search/src/details/DTK_DetailsDistributedSearchTreeImpl.hpp

```diff
@@ -0,0 +1,535 @@
+        Teuchos::ArrayView<int>( export_ranks.data(), n_exports ) );
+
+    // export_ranks already has adequate size since it was used as a buffer to
+    // make the new communication plan.
+    fill( export_ranks, comm_rank );
+
+    Kokkos::View<int *, DeviceType> export_ids( ids.label(), n_exports );
+    Kokkos::parallel_for(
+        REGION_NAME( "fill_buffer" ),
+        Kokkos::RangePolicy<ExecutionSpace>( 0, n_fwd_queries ),
+        KOKKOS_LAMBDA( int q ) {
+            for ( int i = offset( q ); i < offset( q + 1 ); ++i )
+            {
+                export_ids( i ) = ids( q );
+            }
+        } );
+    Kokkos::fence();
+    Kokkos::View<int *, DeviceType> export_indices = indices;
+
+    Kokkos::View<int *, DeviceType> import_indices( indices.label(),
+                                                    n_imports );
+    Kokkos::View<int *, DeviceType> import_ranks( ranks.label(), n_imports );
+    Kokkos::View<int *, DeviceType> import_ids( ids.label(), n_imports );
+    send_across_network( distributor, export_indices, import_indices );
+    send_across_network( distributor, export_ranks, import_ranks );
+    send_across_network( distributor, export_ids, import_ids );
+
+    ids = import_ids;
+    ranks = import_ranks;
+    indices = import_indices;
+
+    //\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\//
+    //\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\//
+    if ( distances_ptr )
+    {
+        Kokkos::View<double *, DeviceType> &distances = *distances_ptr;
+        Kokkos::View<double *, DeviceType> export_distances = distances;
+        Kokkos::View<double *, DeviceType> import_distances( distances.label(),
+                                                             n_imports );
+        send_across_network( distributor, export_distances, import_distances );
+        distances = import_distances;
+    }
+    //\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\//
+}
+
+template <typename DeviceType>
+template <typename Query>
+void DistributedSearchTreeImpl<DeviceType>::filter_results(
+    Kokkos::View<Query *, DeviceType> queries,
+    Kokkos::View<double *, DeviceType> distances,
+    Kokkos::View<int *, DeviceType> &indices,
+    Kokkos::View<int *, DeviceType> &offset,
+    Kokkos::View<int *, DeviceType> &ranks )
+{
+    int const n_queries = queries.extent_int( 0 );
+    // truncated views are prefixed with an underscore
+    Kokkos::View<int *, DeviceType> _offset( offset.label(), n_queries + 1 );
+    fill( _offset, 0 );
+
+    Kokkos::parallel_for( REGION_NAME( "count_results_per_query" ),
+                          Kokkos::RangePolicy<ExecutionSpace>( 0, n_queries ),
+                          KOKKOS_LAMBDA( int q ) {
+                              _offset( q ) = KokkosHelpers::min(
+                                  offset( q + 1 ) - offset( q ),
+                                  queries( q )._k );
```

Comment:
Either we found more than k neighbors in which case we truncate ( found > k -> k )
Or we found less than k and we keep all our results ( found < k - > found )

Does that make sense?